javascript/implementation/Keyed.ts

```diff
@@ -79,6 +85,57 @@ export class Keyed<
         return result !== undefined;
     }
 
+    async reset(
+        toTime?: AsOf,
+        bundlerOrComment?: Bundler | string
+    ): Promise<void> {
+        if (!toTime) {
+            // If no time is specified, we are resetting to epoch, which is just a clear
+            this.clear(false, bundlerOrComment);
+        } else {
+            let bundler: Bundler;
+            let immediate = true;
+            if (typeof bundlerOrComment === "string") {
+                bundler = new Bundler(bundlerOrComment);
+            } else if (bundlerOrComment instanceof Bundler) {
+                immediate = false;
+                bundler = bundlerOrComment;
+            } else {
+                bundler = new Bundler();
+            }
+            const thenMap = await this.toMap(toTime);
+            const nowMap = await this.toMap();
+            // For the entries we currently have, either delete them or update their value
+            for (const [keyNow, valueNow] of nowMap.entries()) {
+                const genericKeyNow = this.storageKeyToGeneric(keyNow);
+
+                const valueThen = thenMap.get(keyNow);
+                if (valueThen !== valueNow) {
+                    if (valueThen === undefined) {
+                        await this.addEntry(
+                            genericKeyNow,
+                            Container.DELETION,
+                            bundler
+                        );
+                    } else {
+                        await this.addEntry(genericKeyNow, valueThen, bundler);
+                        // Done processing this key in the thenMap
+                        ensure(thenMap.delete(keyNow), "failed to delete?");
+                    }
+                }
+            }
+            // Now add new entries for those that were there at toTime, but not now
+            // thenMap will be smaller since the entries there now have been removed.
+            for (const [keyThen, valueThen] of thenMap.entries()) {
+                const genericKeyThen = this.storageKeyToGeneric(keyThen);
+                await this.addEntry(genericKeyThen, valueThen, bundler);
```

Comment:
I'm not in love with the way I've done this, but I don't know another way. I'd love to figure out how to do it in one loop if possible.

The dilemma is: if I loop through the entries there now, we can compare them to what used to be there. However, if an entry was deleted now, but existed then, it would not be in the current entries, and would not be revived. Same problem looping through then entries, just the other way around.

Let me know if I am missing something obvious, or if there just isn't a better way.